File: 0153/binary_search.cpp
#include <cassert>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
 public:
  int findMin(vector<int>& nums) {
    int start = 0;
    int mid = nums.size() / 2;
    int end = nums.size() - 1;
    int min = nums[0];
    while (start < end) {
      if (nums[start] < nums[end]) {
        min = nums[start];
        break;
      }
      mid = (end + start) / 2;
      if (nums[start] > nums[mid]) {
        min = nums[mid];
        end = mid;
        continue;
      }
      if (nums[mid] > nums[end]) {
        min = nums[end];
        start = mid + 1;
        continue;
      }
    }
    return min;
  }
};

int main() {
  Solution sol;
  vector<int> nums;
  nums = {1};
  assert(sol.findMin(nums) == 1);

  nums = {1, 2};
  assert(sol.findMin(nums) == 1);

  nums = {2, 1};
  assert(sol.findMin(nums) == 1);

  nums = {1, 2, 3, 4};
  assert(sol.findMin(nums) == 1);

  nums = {4, 1, 2, 3};
  assert(sol.findMin(nums) == 1);

  nums = {3, 4, 1, 2};
  assert(sol.findMin(nums) == 1);

  nums = {2, 3, 4, 1};
  assert(sol.findMin(nums) == 1);

  nums = {3, 4, 5, 1, 2};
  assert(sol.findMin(nums) == 1);

  nums = {4, 5, 6, 7, 0, 1, 2};
  assert(sol.findMin(nums) == 0);

  nums = {2, 3, 4, 5, 6, 7, 8, 9, 1};
  assert(sol.findMin(nums) == 1);
}

File: 0154/binary_search.py
from typing import List


class Solution:
    def findMin(self, nums: List[int]) -> int:
        left, right = 0, len(nums) - 1
        while left + 1 < right:
            mid = int((right - left) / 2) + left
            # pivot in the left half
            if nums[left] > nums[mid]:
                right = mid
            # pivot is in the right half
            elif nums[mid] > nums[right]:
                left = mid
            # no pivot is in this search space
            elif nums[left] < nums[mid] and nums[mid] < nums[right]:
                break
            # when nums[left] == nums[mid] or nums[mid] == nums[right],
            else:
                right -= 1
        return min(nums[left], nums[right])


if __name__ == '__main__':
    sol = Solution()

    nums, expected = [3, 4, 5, 1, 2], 1
    assert sol.findMin(nums) == expected

    nums, expected = [4, 5, 6, 7, 0, 1, 2], 0
    assert sol.findMin(nums) == expected

    nums, expected = [1, 2, 3], 1
    assert sol.findMin(nums) == expected

    nums, expected = [3, 3, 1, 3], 1
    assert sol.findMin(nums) == expected
